### PR TITLE
pkg simpleschema: adds SimpleSchema enum support

### LINKED ISSUE #304: Implement SimpleSchema enum support
### Feature Description

Currently the `enum` marker type is not implemented in SimpleSchema, causing validation errors when users try to use it:
```
failed to parse field schema for mode: invalid marker key 'enum': unknown marker type: enum
```

Example usage that fails:
```yaml
mode: string | enum="debug,info,warn,error" default="info"
```

The implementation can start here, where the enum marker is defined but not implemented yet: https://github.com/kro-run/kro/blob/fe885014289a7cbad1eaff76d7e0109414f1412b/pkg/simpleschema/markers.go#L40-L49

### PR DESCRIPTION
**Pull request by DhairyaMajmudar**
## Description

1. Adds Enum Support to pkg simpleschema
2. Adds relevant tests cases for the same

fixes #304

### CODE REVIEW

File: pkg/simpleschema/markers.go [hunk 1]
```diff
@@ -50,12 +50,15 @@ const (
 	MarkerTypeMinimum MarkerType = "minimum"
 	// MarkerTypeMaximum represents the `maximum` marker.
 	MarkerTypeMaximum MarkerType = "maximum"
+
```
Comment (a-hilaly) [inline pkg/simpleschema/markers.go]:
```suggestion
```

File: pkg/simpleschema/transform.go [hunk 2]
```diff
@@ -212,6 +213,45 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+				case "boolean":
+					if val == "true" || val == "false" {
+						rawValue = []byte(val)
+						enumJsonValues = append(enumJsonValues, extv1.JSON{Raw: rawValue})
+						continue
+					}
+				default:
+					rawValue = []byte(val)
+					enumJsonValues = append(enumJsonValues, extv1.JSON{Raw: rawValue})
+				}
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go]: Not sure if we want to support JSON and boolean enums

File: pkg/simpleschema/transform.go [hunk 3]
```diff
@@ -212,6 +213,45 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+				switch schema.Type {
+				case "string":
+					rawValue = []byte(fmt.Sprintf("\"%s\"", val))
+				case "integer":
+					if _, err := strconv.ParseInt(val, 10, 64); err == nil {
+						rawValue = []byte(val)
+						enumJsonValues = append(enumJsonValues, extv1.JSON{Raw: rawValue})
+						continue
+					}
+				case "number":
```
Comment (matthchr) [inline pkg/simpleschema/transform.go]: This case is missing a test

File: pkg/simpleschema/transform.go [hunk 4]
```diff
@@ -212,6 +213,45 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
 			if val, err := strconv.ParseFloat(marker.Value, 64); err == nil {
 				schema.Maximum = &val
 			}
+		case MarkerTypeEnum:
+			enumValues := strings.Split(marker.Value, ",")
+
+			var enumJsonValues []extv1.JSON
```
Comment (matthchr) [inline pkg/simpleschema/transform.go]: minor: 
```suggestion
			var enumJSONValues []extv1.JSON
```

File: pkg/simpleschema/transform.go [hunk 5]
```diff
@@ -212,6 +213,45 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+
+			for _, val := range enumValues {
+				val = strings.TrimSpace(val)
+				var rawValue []byte
+
+				switch schema.Type {
+				case "string":
+					rawValue = []byte(fmt.Sprintf("\"%s\"", val))
+				case "integer":
+					if _, err := strconv.ParseInt(val, 10, 64); err == nil {
```
Comment (matthchr) [inline pkg/simpleschema/transform.go]: We should do something if err is not nil - probably amend this function to return an error and then handle/return that error in `parseFieldSchema`?

File: pkg/simpleschema/transform.go [hunk 6]
```diff
@@ -212,6 +213,45 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+						enumJsonValues = append(enumJsonValues, extv1.JSON{Raw: rawValue})
+						continue
+					}
+				case "number":
+					if _, err := strconv.ParseInt(val, 10, 64); err == nil {
+						rawValue = []byte(val)
+						enumJsonValues = append(enumJsonValues, extv1.JSON{Raw: rawValue})
+						continue
+					}
+				case "boolean":
```
Comment (matthchr) [inline pkg/simpleschema/transform.go]: IMO boolean enums are nonsensical - remove?

**Review by a-hilaly — COMMENTED** [review]
Thanks @DhairyaMajmudar ! i have two nits below:

File: pkg/simpleschema/transform_test.go [hunk 7]
```diff
@@ -279,6 +279,64 @@ func TestBuildOpenAPISchema(t *testing.T) {
 			},
 			wantErr: false,
 		},
+		{
+			name: "Schema with multiple enum types",
+			obj: map[string]interface{}{
+				"logLevel": "string | enum=\"debug,info,warn,error\" default=\"info\"",
+				"features": map[string]interface{}{
+					"debugMode": "boolean | enum=\"true,false\" default=false",
+					"logFormat": "string | enum=\"json,text,csv\" default=\"json\"",
```
Comment (matthchr) [inline pkg/simpleschema/transform_test.go:287]: I think you're missing some negative test cases where users pass invalid values to the enum - i.e. it's an int but they pass strings, or its a string and they pass a totally empty value (`,,`)

File: pkg/simpleschema/transform.go (hunk 2 above)
Comment (matthchr) [reply pkg/simpleschema/transform.go]: Agree, IMO no boolean/json enums.

Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Removed in the commit ee57ea0783a0c90936fb49bd69adde1bd98df7f1

File: pkg/simpleschema/transform.go (hunk 5 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Modified the `parseFieldSchema` and `applyMarkers`  with proper error handling in the commit  5db9e88ba900b9682fe647184dafc4fb7ea012ae

File: pkg/simpleschema/transform.go (hunk 3 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Added in the commit 74fd90c0c87f18cdc0272db0bf758b71b517d3e3

File: pkg/simpleschema/transform_test.go (hunk 7 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform_test.go:287]: Done in the commit 763c707d013ee6b4b35bef01041131c8baf1980e

**Comment by DhairyaMajmudar** [comment]
Thank you @a-hilaly @matthchr for the review, updated my PR following the same.

File: pkg/simpleschema/transform.go [hunk 8]
```diff
@@ -205,15 +208,52 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+				case "string":
+					rawValue = []byte(fmt.Sprintf("\"%s\"", val))
+				case "integer":
+					if _, err := strconv.ParseInt(val, 10, 64); err != nil {
+						return fmt.Errorf("failed to parse integer enum value: %w", err)
+					} else {
+						rawValue = []byte(val)
+					}
+				case "number":
+					if _, err := strconv.ParseFloat(val, 64); err != nil {
```
Comment (matthchr) [inline pkg/simpleschema/transform.go]: FWIW there is a general guidance to avoid floats in k8s APIs.

I am not sure if we want to support float enums - support floats (which we already support AFAIK) is one thing, but a float enum is a pretty weird/niche case and I'm wondering if we just want to drop this for now, at least until some user has a concrete ask/reason for why they might want something like this. I'm not sure I've ever seen a float enum in the wild.

@a-hilaly wdyt?

File: pkg/simpleschema/transform_test.go [hunk 9]
```diff
@@ -279,6 +279,73 @@ func TestBuildOpenAPISchema(t *testing.T) {
...
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "invalid enum type",
```
Comment (matthchr) [inline pkg/simpleschema/transform_test.go:332]: one more test for empty enum value?

`"threshold": "int | enum=\"1,,3\"",` - I think we expect error here too, no?

Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform_test.go:332]: Yeah that one is required, Thanks for suggestion. Added one 👍🏻

Comment (matthchr) [reply pkg/simpleschema/transform_test.go:332]: I think this needs to be in a separate test because otherwise the `threshold` will alawys fail. For failure testscases you always need 1 failure-type per test.

File: pkg/simpleschema/transform.go (hunk 8 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Attaching some references for why not to use floats

1. https://github.com/zecke/Kubernetes/blob/master/docs/devel/api-conventions.md#primitive-types
2. https://github.com/kubernetes-sigs/controller-tools/issues/245

POV: We should remove the case for number as these references tells not support floating values.

File: pkg/simpleschema/transform_test.go (hunk 9 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform_test.go:332]: Got you point, added both of them into one as they are same outputs `nil` 😁

NVM separated both of them now.

File: pkg/simpleschema/transform.go (hunk 8 above)
Comment (a-hilaly) [reply pkg/simpleschema/transform.go]: Agreed, we can drop support for enum floats. We can remove the `number` case

File: pkg/simpleschema/transform.go [hunk 10]
```diff
@@ -205,15 +208,52 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+		case MarkerTypeEnum:
+			enumValues := strings.Split(marker.Value, ",")
+
+			var enumJSONValues []extv1.JSON
+
+			for _, val := range enumValues {
+				val = strings.TrimSpace(val)
+				var rawValue []byte
+
+				switch schema.Type {
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go:229]: Shall we just keep `string` and `integer` here? for the default we can just error saying that enums are only supported for string and integer types

File: pkg/simpleschema/atomic.go [hunk 11]
```diff
@@ -34,11 +34,13 @@ const (
...
 	// AtomicTypeString represents a string value.
 	AtomicTypeString AtomicType = "string"
+	// AtomicTypeNumber represents a number value.
+	AtomicTypeNumber AtomicType = "number"
 )
 
 func isAtomicType(s string) bool {
 	switch AtomicType(s) {
-	case AtomicTypeBool, AtomicTypeInteger, AtomicTypeFloat, AtomicTypeString:
+	case AtomicTypeBool, AtomicTypeInteger, AtomicTypeFloat, AtomicTypeString, AtomicTypeNumber:
```
Comment (a-hilaly) [inline pkg/simpleschema/atomic.go]: Can introduce this change in a separate PR instead? feels like it's addressing a different problem and will very likely be a breaking change.

Comment (a-hilaly) [reply pkg/simpleschema/atomic.go]: Now that i'm thinking about it we might have some miss leading documentation here https://kro.run/docs/concepts/simple-schema - it should be `float` not `number`, so maybe not that breaking of a change :)

Comment (DhairyaMajmudar) [reply pkg/simpleschema/atomic.go]: This won't be required now, as it was associated with the number case in `transform.go` file, but as we are dropping it I've undoed these changes.

File: pkg/simpleschema/transform.go (hunk 8 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Done, removed the number cases.

File: pkg/simpleschema/transform.go [hunk 12]
```diff
@@ -205,15 +208,46 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+				case "string":
+					rawValue = []byte(fmt.Sprintf("\"%s\"", val))
+				case "integer":
+					if _, err := strconv.ParseInt(val, 10, 64); err != nil {
+						return fmt.Errorf("failed to parse integer enum value: %w", err)
+					} else {
+						rawValue = []byte(val)
+					}
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go]: nit: `else` is not needed here :)
```suggestion
					}
					rawValue = []byte(val)
```

File: pkg/simpleschema/transform.go [hunk 13]
```diff
@@ -205,15 +208,46 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+			if val, err := strconv.ParseFloat(marker.Value, 64); err != nil {
+				return fmt.Errorf("failed to parse minimum enum value: %w", err)
+			} else {
 				schema.Minimum = &val
 			}
 		case MarkerTypeMaximum:
-			if val, err := strconv.ParseFloat(marker.Value, 64); err == nil {
+			if val, err := strconv.ParseFloat(marker.Value, 64); err != nil {
+				return fmt.Errorf("failed to parse maximum enum value: %w", err)
+			} else {
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go]: two nits:
- We can have this in a seperate PR since it's not related to enums
- We don't need an `else` statement

File: pkg/simpleschema/transform.go [hunk 14]
```diff
@@ -205,15 +208,46 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
 				schema.Maximum = &val
 			}
+		case MarkerTypeEnum:
+			enumValues := strings.Split(marker.Value, ",")
+
+			var enumJSONValues []extv1.JSON
+
+			for _, val := range enumValues {
+				val = strings.TrimSpace(val)
+				var rawValue []byte
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go]: Not sure how is this handling invalid enums such as `a,b,,,,c` - and i'm not quiet grokking how are unit tests passing here. Could be something wrong with the assert statements

File: pkg/simpleschema/transform.go (hunk 13 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Gotcha, let me remove this from here and segregate into an another PR

File: pkg/simpleschema/transform.go (hunk 14 above)
Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: There's really a great edge case as in this case `string.Split(marker.Value, ",")` will make it `["a", "b", "", "", "", "c"]` and later loop and code further make this an empty string which is wrong. As in the case of integers it fails early when it hits the empty string and my test case checks the string that's why its failing. Let me add a if statement and a test case for this.

Comment (a-hilaly) [reply pkg/simpleschema/transform.go]: Good catch, indeed `strconv.ParseInt` errors with empty strings

File: pkg/simpleschema/transform.go [hunk 15]
```diff
@@ -212,8 +215,39 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+
+				if val == "" {
+					return fmt.Errorf("empty enum values are not allowed")
+				}
+
+				var rawValue []byte
+
+				switch schema.Type {
+				case "string":
+					rawValue = []byte(fmt.Sprintf("\"%s\"", val))
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go]: nit: you can use %q to quote-format a string 
```suggestion
					rawValue = []byte(fmt.Sprintf("%q", val))
```

File: pkg/simpleschema/transform_test.go [hunk 16]
```diff
@@ -279,6 +279,79 @@ func TestBuildOpenAPISchema(t *testing.T) {
...
+			obj: map[string]interface{}{
+				"threshold": "integer | enum=\"1,2,three\"",
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "empty enum integer type",
```
Comment (a-hilaly) [inline pkg/simpleschema/transform_test.go]:
```suggestion
			name: "Invalid integer enum marker",
```

File: pkg/simpleschema/transform_test.go [hunk 17]
```diff
@@ -279,6 +279,79 @@ func TestBuildOpenAPISchema(t *testing.T) {
...
+			obj: map[string]interface{}{
+				"errorCode": "integer | enum=\"1,,3\"",
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "string enum with empty values",
```
Comment (a-hilaly) [inline pkg/simpleschema/transform_test.go]:
```suggestion
			name: "invalid string enum marker",
```

File: pkg/simpleschema/transform.go [hunk 18]
```diff
@@ -212,8 +215,39 @@ func (tf *transformer) applyMarkers(schema *extv1.JSONSchemaProps, markers []*Ma
...
+			for _, val := range enumValues {
+				val = strings.TrimSpace(val)
+
+				if val == "" {
+					return fmt.Errorf("empty enum values are not allowed")
+				}
+
+				var rawValue []byte
+
+				switch schema.Type {
```
Comment (a-hilaly) [inline pkg/simpleschema/transform.go]: nit: while not specific to Go, it's a common practice to group related statements more closely together with fewer blank lines between them. Typically you'll see blank lines only to separate logical "paragraphs" of code
```suggestion
			var enumJSONValues []extv1.JSON

			enumValues := strings.Split(marker.Value, ",")
			for _, val := range enumValues {
				val = strings.TrimSpace(val)
				if val == "" {
					return fmt.Errorf("empty enum values are not allowed")
				}
				
				var rawValue []byte
				switch schema.Type {
```

Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform.go]: Done with all suggestions 👍🏻

File: pkg/simpleschema/transform_test.go [hunk 19]
```diff
@@ -279,6 +279,79 @@ func TestBuildOpenAPISchema(t *testing.T) {
...
+			wantErr: true,
+		},
+		{
+			name: "Invalid integer enum marker",
+			obj: map[string]interface{}{
+				"errorCode": "integer | enum=\"1,,3\"",
+			},
+			want:    nil,
+			wantErr: true,
+		},
```
Comment (a-hilaly) [inline pkg/simpleschema/transform_test.go]: Shall we split this into two test cases? one testing that `1,,3` is invalid asserting `empty enum values are not allowed` and another testing that `1,2,3,abc` is invalid asserting `invalid syntax` (which i believe is the error returned by `strconv.ParseInt`)

Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform_test.go]: Yep, let's do it

Comment (DhairyaMajmudar) [reply pkg/simpleschema/transform_test.go]: Done 👍🏻

**Review by a-hilaly — APPROVED** [review]
Thanks a lot @DhairyaMajmudar !

**Review by matthchr — APPROVED** [review]
